### DISCUSSION TITLE
How to use Vapor Middleware

### DISCUSSION THREAD

**Discussion by saroar** [pointfreeco/vapor-routing#16] [Q&A]
<img width="780" alt="Screenshot 2022-07-21 at 11 38 50" src="https://user-images.githubusercontent.com/8770772/180177094-6269fb29-e62f-4be2-abb6-7682a2859f4c.png">
how do i can use it ? is there any option?

**Comment by saroar** [comment]
```swift
import Vapor
import JWT
import MongoKitten

public final class JWTMiddleware: Middleware {
    public init() {}

    public func respond(to request: Request, chainingTo next: Responder) -> EventLoopFuture<Response> {

        guard let token = request.headers.bearerAuthorization?.token.utf8 else {
            return request.eventLoop.makeFailedFuture(
                Abort(.unauthorized, reason: "Missing authorization bearer header")
            )
        }

        do {
            request.payload = try request.jwt.verify(Array(token), as: Payload.self)
        } catch let JWTError.claimVerificationFailure(name: name, reason: reason) {
            request.logger.error("JWT Verification Failure: \(name) \(reason)")
            return request.eventLoop.makeFailedFuture(JWTError.claimVerificationFailure(name: name, reason: reason))
        } catch let error {
            return request.eventLoop.makeFailedFuture(
                Abort(.unauthorized, reason: "You are not authorized this token \(error)")
            )
        }

        return next.respond(to: request)
    }
}

extension AnyHashable {
    static let payload: String = "jwt_payload"
}

extension Request {
    var loggedIn: Bool {
        return self.storage[PayloadKey.self] != nil ?  true : false
    }

    var payload: Payload {
        get { self.storage[PayloadKey.self]! } // should not use it
        set { self.storage[PayloadKey.self] = newValue }
    }
}
```

**Comment by saroar** [comment]
```swift
my routes 
try app.group("v1") { api in
    let events = api.grouped("events")
    let eventsAuth = events.grouped(JWTMiddleware())
    try eventsAuth.register(collection: EventController() )
}
// inside controller i use like this 
if request.loggedIn == false { throw Abort(.unauthorized) }
```
but for vapor-router no option

**Comment by stephencelis** [comment]
Hi @saroar! Since this is a question and not a bug report, I'm going to convert this to a discussion where more eyes might see it. If there is a deficiency in the library we'd definitely take a PR to correct it, though!

**Comment by saroar** [comment]
I really just loving it and enjoy how awesome it just working but when I send header from client I have to get it in my server and check it 
this way 
`JWTMiddleware` I already posted code in up please check but which not working with this package 

<img width="602" alt="Screenshot 2022-08-15 at 22 06 04" src="https://user-images.githubusercontent.com/8770772/184700237-c38eee56-45e6-4cf0-b828-2c331a7a5934.png">

this is what we have discuss on discord with @0xTim
<img width="961" alt="Screenshot 2022-08-15 at 22 08 31" src="https://user-images.githubusercontent.com/8770772/184700619-128c70b5-e749-4889-9b0f-a27b15e2e711.png">

**Comment by finestructure** [comment]
What a coincidence, I'm just in the process of converting our routing to use the parser printer and I'm also running into the issue that I've got one route wrapped in auth middlewares:

```swift
        app.group(User.TokenAuthenticator(), User.guardMiddleware()) { protected in
            protected.on(.POST, SiteURL.api(.versions(.key, .builds)).pathComponents,
                         use: API.BuildController.create)
        }

```

`SiteURL.api(.versions(.key, .builds)).pathComponents` is our existing, less-capable variant of consolidated routing and it's not clear to me how I would attach the auth middlewares to a single (or a subset) of parser-printer routes via the

```swift
    app.mount(SiteRoute.router, use: SiteRoute.handler)
```

API.

I suppose one way could be a custom `mount` overload with the additional middlewares in it and then to define a new parser-printer router that handles just that protected route. But that's probably not going to work as is as because it would fail for all the routes it _doesn't_ handle 🤔

**Reply by stephencelis** [reply]
We're down to support more nuanced routing/middleware like this, but aren't quite sure what would meet the typical Vapor user's needs. It seems like there maybe should be a way to inject middleware after `swift-url-routing` has done its job, but we're not quite sure how to approach that.

@finestructure Does that kind of thing even sound feasible?

**Reply by finestructure** [reply]
> but aren't quite sure what would meet the typical Vapor user's needs

I think the canonical use-case is wrapping certain routes with authentication middlewares, like above. This _has_ to come up with practically any Vapor app that does more than vending public endpoints.

I've tried to sketch out something that might work with relatively minor changes. Starting from the following `SiteRoute` declaration [as I've set it up here](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/pull/1973/files#diff-bcf3eabb631e7d33aeae647ed5ea186d3a2ab93683ffb0aba0c6959f9a593f6d):

```swift
// route + handler
enum SiteRoute {
    static let route = OneOf { ... }
    static func handler(req: Request, route: SiteRoute) async throws -> AsyncResponseEncodable { ... }
}

// mount
app.mount(SiteRoute.router, use: SiteRoute.handler)
```

What I'd need is something like this instead:

```swift
// public routes
enum PublicSiteRoute { ... }

// protected routes (to be wrapped in auth middlewares)
enum ProtectedSiteRoute { ... }
```

I (think) can't just do this

```swift
app.mount(PublicSiteRoute.router, use: PublicSiteRoute.handler)
// imagine there was a "mount" overload that wrapped in additional middlewares
app.mount(
	ProtectedSiteRoute.router,
	use: ProtectedSiteRoute.handler, 
	middlewares: User.TokenAuthenticator(), User.guardMiddleware()
)
```

because the routers are exclusive - they expect to handle all routes. There'd have to be a way to chain them together such that only if they all fail does the routing fail.

I wonder if there's a way to write a mount that combines everything and allows providing mappings against middlewares. Something like

```swift
public struct MountPoint<R: Parser> where R.Input == URLRequestData {
    var router: R
    var handler: (Request, R.Output) async throws -> AsyncResponseEncodable
    var middlewares: [Middleware] = []
}

app.mount(
	.init(router: PublicSiteRoute, handler: PublicSiteRoute.handler),
	.init(router: ProtectedSiteRoute, handler: ProtectedSiteRoute.handler, middlewares: [User.TokenAuthenticator(), User.guardMiddleware()])
)
```

And then mount does some work to chain things together:

```swift
extension Application {
    public func mount<R1: Parser, R2: Parser>(
        _ mp1: MountPoint<R1>,
        _ mp2: MountPoint<R2>
    ) where R1.Input == URLRequestData, R2.Input == URLRequestData {
        self.middleware.use(AsyncChainedRoutingMiddleware(mountPoint1: mp1, mountPoint2: mp2))
    }
}

public struct AsyncChainedRoutingMiddleware<R1: Parser, R2: Parser>: AsyncMiddleware
where R1.Input == URLRequestData, R2.Input == URLRequestData
{
    var mountPoint1: MountPoint<R1>
    var mountPoint2: MountPoint<R2>

    public func respond(to request: Vapor.Request, chainingTo next: Vapor.AsyncResponder) async throws -> Vapor.Response {
        do {
            // try first router + wrap in its middlerwares
        } catch let routingError {
            // try second router + wrap in its middlewares
        }
    }
}
```

Feels a bit clunkier API-wise than simply being able to pick out which routes get wrapped in a middleware in current Vapor though but at least it'd support the use-case... 🤔

Also, this would need a couple higher arity overloads. I don't think there's a way to write that generically across different `Router`s with current Swift. Or if there is, I certainly can't 😆

**Reply by finestructure** [reply]
This API is also not ideal because it breaks composability to a degree. For instance, I have API routes (`/api/...`) that are public and some that are protected. So I'd need to repeat the `api` case with its prefix parsing in both the `Public...` and the `Protected...` routes.

```swift
enum PublicSiteRoute {
    case api(PublicApiRoute)
    ...
    static let router = OneOf {
        // /api/...
        Route(.case(Self.api)) { Path { "api" }; PublicApiRoute.router }
        ...
}
```

```swift
enum ProtectedSiteRoute {
    case api(ProtectedApiRoute)
    ...
    static let router = OneOf {
        // /api/...
        Route(.case(Self.api)) { Path { "api" }; ProtectedApiRoute.router }
        ...
}
```

It would be better if there was a way to attach a middleware in the router somehow.

```swift
enum SiteRoute {
    case api(ApiRoute)
    ...
    static let router = OneOf {
        // /api/...
        Route(.case(Self.api)) { Path { "api" }; ApiRoute.router }
        ...

let auth: [Middleware] = [User.TokenAuthenticator(), User.guardMiddleware()]

enum ApiRoute {
    case version
    case build

    static let router = OneOf {
        // GET /version
        Route(.case(Self.version)) { Path { "version" } }

        // POST /build
        Route(.case(self.build), middlewares: auth) { ... }
    }
}
```

But that's a very disruptive change, because `Parser` doesn't know about any additional middlewares and `mount`/`AsyncRoutingMiddleware` wouldn't be able to access it. I guess there'd need to be a whole new type that'd need to be passed through here.

**Reply by saroar** [reply]
@stephencelis can you help please I need to handle two type of Middleware how I can do it for this I am having very big issue now I cant go farther with my project I am stuck with more then month :( 

for now I am doing like this but I cant do it

```swift
public final class JWTMiddleware: AsyncMiddleware {
    public init() {}

    public func respond(to req: Request, chainingTo next: AsyncResponder) async throws -> Response {
        guard let token = req.headers.bearerAuthorization?.token.utf8 else {

->`this part for is public and some of request came from web api for admin`
            if [
                "/v1/auth/login",
                "/v1/auth/verify_sms",
                "/v1/devices",
                "/v1/words",
                "/terms",
                "/privacy",
                "/words",
                "/auth/login",
                "/auth/register",
                "/auth/create",
                "/admin",
                "/terms",
                "/privacy",
                "/"
            ].contains(req.url.string) {
                return try await next.respond(to: req)
            }

            return Response(status: .unauthorized, body: .init(string: "Missing authorization bearer header"))
        }

        do {
            req.payload = try req.jwt.verify(Array(token), as: Payload.self)
        } catch let JWTError.claimVerificationFailure(name: name, reason: reason) {
            throw JWTError.claimVerificationFailure(name: name, reason: reason)
        } catch let error {
            debugPrint("\(self) \(#line) \(#file) \(error)")
            return Response(status: .unauthorized, body: .init(string: "You are not authorized this token \(error)"))
        }

        return try await next.respond(to: req)
    }
}
```

**Reply by saroar** [reply]
for example I can by pass `http://192.168.1.29:7070/api/auth/email-verification?token=ed66f8954c66d2a281f41fa8e35e9a1bf903fdb9067cc9979b79f4d82af42556` :(

**Reply by nevillco** [reply]
I got this working by extending `app.mount` to take an additional closure of `(Route) -> Migration?`, so you can define a migration to run according to the route, right before the default, consuming behavior of `app.mount`. I'm not necessarily sure it's the best API either, but it is working for the use case of having auth middleware around certain routes but not others. Diff here: https://github.com/pointfreeco/vapor-routing/compare/main...nevillco:vapor-routing:main
